### DISCUSSION TITLE
技術スタック

### DISCUSSION THREAD

**Discussion by sanoyo** [sanoyo/ultrachat#4] [General]
# フロントエンド
- reasct native
- WebSockets
- WebRTC

# バックエンド
- golang
- WebSockets

# インフラ
- aws
- terraform

**Comment by sanoyo** [comment]
とりあえず方向性(要求、要件)を決めたい

- UI
- 機能

**Reply by sanoyo** [reply]
ここに書いていく

**Reply by sanoyo** [reply]
やりたい技術スタックで構成する

**Reply by sanoyo** [reply]
やりたい技術スタック
- WebSockets
- WebRTC
- graphql
- ecs or [これ](https://github.com/aws-samples/aws-bookstore-demo-app)
- [weaver](https://github.com/ServiceWeaver/weaver)

**Reply by ddddddO** [reply]
やりたいこと
- react native
- graphql

**Reply by sanoyo** [reply]
下記でどう？

- WebSockets
- WebRTC
- graphql
- ecs
- [weaver](https://github.com/ServiceWeaver/weaver)

**Reply by sanoyo** [reply]
これで決まり

**Reply by ddddddO** [reply]
React Native x **Expo** でいく！
ちょっと前に試してみて、開発進めやすそうではと思った
ref: https://github.com/ddddddO/BookBark#readme

**Reply by ddddddO** [reply]
- [AppSync](https://aws.amazon.com/jp/appsync/)
  - [「リアルタイムのコラボレーションおよびチャットアプリケーションを構築する」](https://aws.amazon.com/jp/blogs/mobile/aws-appsync-real-time-reference-architecture/)
- [【備考】Subscriptionについて](https://zenn.dev/eringiv3/books/a85174531fd56a/viewer/d44baf)
- [GraphQLにおけるSubscription処理について(実装例: Amplify + AppSync)](https://qiita.com/yoshii0110/items/3d9ec03215537646b65c)
- [apollo-clientをReactNative(Expo)で](https://zenn.dev/knaka0209/books/cd1518009314d2/viewer/c90eca)
  - [あなたのプロダクトに Apollo Client は必要ないかもしれない](https://user-first.ikyu.co.jp/entry/2022/07/01/121325)
- [AppSync & GraphQL 入門](https://qiita.com/umamichi/items/43e8ca7c86e937e4bf54)

**Reply by ddddddO** [reply]
- GraphQL/gplgen/subscribe(local動作確認用)
  - [複雑でない実装](https://www.ohitori.fun/entry/graphql-subscription-in-go)
  - [Redis使った本格派](https://kaminashi-developer.hatenablog.jp/entry/2020/12/11/093000)

**Comment by sanoyo** [comment]
# インフラ

以下を踏まえ、アーキテクチャを考える

- WebSockets
- WebRTC
- graphql
- ecs
- [weaver](https://github.com/ServiceWeaver/weaver)

**Reply by sanoyo** [reply]
雑だけどこんな感じ

```mermaid
flowchart TB
  client-->alb
  alb-->appasync
  appasync-->ecsservice
  ecsservice-->ecstask
```

**Reply by sanoyo** [reply]
![image](https://user-images.githubusercontent.com/24674237/224470020-2791a669-6c22-4de9-b1cc-8eca841670f3.png)

https://aws.amazon.com/jp/blogs/mobile/aws-appsync-real-time-reference-architecture/

**Comment by sanoyo** [comment]
# バックエンド
- golangとgraphqlの開発
- schemaをresolverが必要
  - これでmockserver作れないか

**Comment by sanoyo** [comment]
# 機能
Discordは、主に以下のような機能

- ボイスチャット機能 - グループや個人間で音声通話を行うことができます。
- テキストチャット機能 - グループや個人間でテキストチャットを行うことができます。
- 画面共有機能 - 画面共有をすることができます。
- ビデオ通話機能 - グループや個人間でビデオ通話を行うことができます。
- サーバー機能 - チャットルームをまとめたサーバーを作成し、複数人で情報共有やコミュニケーションを行うことができます。
- チャンネル機能 - サーバー内で複数のチャットルームを作成することができます。
- ロール機能 - サーバー内のユーザーやチャンネルに権限を付与することができます。
- 絵文字機能 - カスタムの絵文字を作成し、チャットやリアクションに使用することができます。
- ボット機能 - カスタムのボットを作成し、自動応答や定期的な情報の更新などを行うことができます。
- 通知機能 - メンションや特定の単語が入力された場合に通知を受け取ることができます。
- ファイル共有機能 - ファイルをアップロードし、共有することができます。
- アカウント設定機能 - アカウントの情報や設定をカスタマイズすることができます。
- モバイルアプリ機能 - モバイルアプリを通じて、Discordを利用することができます。

**Reply by sanoyo** [reply]
最初は下記機能
- ボイスチャット機能 - グループや個人間で音声通話を行うことができます。
- テキストチャット機能 - グループや個人間でテキストチャットを行うことができます。
- ビデオ通話機能 - グループや個人間でビデオ通話を行うことができます。

**Reply by sanoyo** [reply]
その中でもこれをできるようにする
- テキストチャット機能 - グループや個人間でテキストチャットを行うことができます。

**Reply by sanoyo** [reply]
UIはcss当てずにテキストフィールドとそれを表示するフィールドがるだけ

**Reply by sanoyo** [reply]
- バックエンドが必要なのか
- appasyncのサブスクリプションの後ろのデータソースをどうするか

**Reply by sanoyo** [reply]
>バックエンドが必要なのか

不要

>appasyncのサブスクリプションの後ろのデータソースをどうするか

dynamodbで良さそう。
auroraでも対応できるがauroraの場合、サブスクリプションをトリガーするためにlambdaを用意しないといけないので、dynamodbがよさそう。

```
AuroraはDynamoDBと異なり、データ更新のトリガーとして直接サブスクリプションを利用することはできません。そのため、Auroraを使用する場合は、変更をトリガーとしてLambdaを起動し、そのLambdaがAppSyncのSubscriptionを通じてクライアントに変更を配信する必要があります。
```

**Reply by sanoyo** [reply]
# GraphQL Schema
https://developer.kaizenplatform.com/entry/laco/2018-06-08
![image](https://user-images.githubusercontent.com/24674237/224472679-b51eac97-d3b0-4eda-a045-fa32182f47b1.png)

**Comment by sanoyo** [comment]
appasyncとdynamodbの構成だとバックエンドのリソースは必要なさそう

**Reply by sanoyo** [reply]
あとはデプロイ方法をamplifyとかで考えるとかかなー
@ddddddO

**Comment by sanoyo** [comment]
- [x] ドメイン
  - awsで取得。
- [x] ネットワーク
  - vpc
  - サブネット
  - ルートテーブル
  - ネットワークコネクション
- [x] セキュリティグループの作成
- [ ] ALB
  - ALBとappasync紐づけ
- [ ] route53
- [ ] テストする

**Reply by sanoyo** [reply]
ALBとappsyncは直接紐づけることはできないみたい。
albをappsyncの手前に配置するためには、lambdaを配置するなどをしなければいけない。

**Reply by sanoyo** [reply]
なので、appsyncとlambdaとdynamodb使う想定でいいんじゃなかろーか

**Reply by ddddddO** [reply]
ちょっと見るのが遅くて🙏🙏！

lambdaは、appsyncの後ろで動くイメージでokですか??
albを手前にもってくるの、websocketどうなるんだろうと思ってたので、appsyncが入り口になるの賛成です！

**Reply by sanoyo** [reply]
全然、お遊びでやってるから良いよ！

>lambdaは、appsyncの後ろで動くイメージでokですか??

そうね、ただ一旦は必要なさそうなんで、必要なタイミングで入れるでいいと思う！

>albを手前にもってくるの、websocketどうなるんだろうと思ってたので、appsyncが入り口になるの賛成です！

や！

**Reply by sanoyo** [reply]
とりまおっちーの実装をs3に上げてそこからappsyncをたたく構成が良いと思ってる！
手順まとめ中ー！

**Reply by sanoyo** [reply]
`npx expo export:web`コマンドでビルドしようとするとエラー出るんだけどわかる？
ちょっとissueにしてみるねー1
@ddddddO

**Reply by sanoyo** [reply]
してみたー！
https://github.com/sanoyo/ultrachat/issues/9

**Reply by ddddddO** [reply]
ちょっと見てみます！🙏

**Reply by sanoyo** [reply]
あざまー！
その間に認証作っておくー。土台となりそうなので、早めにつくっておくー
https://github.com/sanoyo/ultrachat/discussions/10
